### PR TITLE
Optional UI rendering

### PR DESCRIPTION
**Pull request by ickshonpe**
# Objective

Make UI rendering optional.

Quite a few people have been experimenting with using Bevy UI for layout and interaction but replacing the rendering with `bevy_prototype_lyon` or whatever. It's awkward to do though as there is no way to disable the existing UI rendering requiring users to create their own custom node bundles and components. Also, you can't replace the UI's shader and a number of other things.

This PR makes the setup and initialization of UI rendering for the RenderApp optional. Then you can do whatever you want by replacing `build_ui_render` with your own function. For instance, one that loads a custom shader.

The UI layout and interaction components are still updated as normal.

## Solution

Add a field `enable_rendering` to `UiPlugin`.
Only call `build_ui_render` and initialize the `UiPipeline` if `enable_rendering` is false.

I thought about implementing a "bevy_ui_render" feature but suspect everything is too tightly coupled atm and it would be very fragile. Similar to the struggles with the "bevy_text" feature but worse. 

---

## Changelog

  `UiPlugin`
  * Added a bool field `enable_rendering`.
  * Only calls `build_ui_render` and initializes the `UiPipeline` if `enable_rendering` is true.

## Migration Guide

`UiPlugin` has a new field `enable_rendering`. If set to false, the UI's rendering systems won't be added to the `RenderApp` and no UI elements will be drawn. The layout and interaction components will still be updated as normal.

### CODE REVIEW

**Review by alice-i-cecile — APPROVED** [review]
> I thought about implementing a "bevy_ui_render" feature but suspect everything is too tightly coupled atm and it would be very fragile. Similar to the struggles with the "bevy_text" feature but worse.

I would prefer this solution, especially in the long run, but the struggles are entirely plausible. We *should* work towards decoupling things here though.

**Comment by ickshonpe** [comment]
> > I thought about implementing a "bevy_ui_render" feature but suspect everything is too tightly coupled atm and it would be very fragile. Similar to the struggles with the "bevy_text" feature but worse.
> 
> I would prefer this solution, especially in the long run, but the struggles are entirely plausible. We _should_ work towards decoupling things here though.

Definitely agree. My preference I think would be to move the UI renderer into its own crate.

**Comment by nicoburns** [comment]
> Similar to the struggles with the "bevy_text" feature but worse.

This could be fixed by running tests without the `bevy_text` feature on CI (which can be done by running `cargo test` in the `bevy_ui` crate dir rather than just from the root).

File: crates/bevy_ui/src/lib.rs [hunk 1]
```diff
@@ -50,7 +50,11 @@ use update::update_clipping_system;
 
 /// The basic plugin for Bevy UI
 #[derive(Default)]
-pub struct UiPlugin;
+pub struct UiPlugin {
+    /// When set to true, UI rendering won't be added to the `RenderApp` and the UI will not be shown.
+    /// The UI layout and interaction components will still be updated as normal.
+    pub disable_ui_rendering: bool,
```
Comment (janhohenheim) [inline crates/bevy_ui/src/lib.rs:56]: I think having a boolean in the negative is a bit weird, semantically. Could you change this to `ui_rendering: bool` with a default of `true`?

File: crates/bevy_ui/src/lib.rs [hunk 2]
```diff
@@ -173,15 +177,19 @@ impl Plugin for UiPlugin {
...
+            crate::render::build_ui_render(app);
+        }
     }
 
     fn finish(&self, app: &mut App) {
-        let render_app = match app.get_sub_app_mut(RenderApp) {
-            Ok(render_app) => render_app,
-            Err(_) => return,
-        };
+        if !self.disable_ui_rendering {
```
Comment (janhohenheim) [inline crates/bevy_ui/src/lib.rs:186]: Nit: I'd prefer this as an early return

**Review by janhohenheim — CHANGES_REQUESTED** [review]

**Review by pablo-lua — APPROVED** [review]

**Comment by alice-i-cecile** [comment]
@ickshonpe just waiting on merge conflicts now :)

**Pull request merged**